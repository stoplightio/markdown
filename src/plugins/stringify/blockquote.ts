import { safeStringify } from '@stoplight/yaml';
import { Handler } from 'mdast-util-to-markdown';
// @ts-expect-error
import blockquote from 'mdast-util-to-markdown/lib/handle/blockquote';

export const blockquoteHandler: Handler = function (node, _, context) {
  const annotations = (node.data?.hProperties || {}) as any;

  const value = blockquote(node, _, context);

  if (Object.keys(annotations).length) {
    return `<!-- ${safeStringify(annotations, { skipInvalid: true }).trim()} -->

${value}`;
  } else {
    return value;
  }
};
