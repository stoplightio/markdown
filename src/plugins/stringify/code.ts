import * as Yaml from '@stoplight/yaml';
import { Handler } from 'mdast-util-to-markdown';
// @ts-expect-error
import code from 'mdast-util-to-markdown/lib/handle/code';

const { safeStringify } = Yaml;

export const codeHandler: Handler = function (node, _, context) {
  const { lang, meta: _meta, ...annotations } = (node.data?.hProperties || {}) as any;

  // if node has resolved prop we want to replace value with resolved value when stringifying
  if (node.resolved) {
    node.value =
      node.lang === 'json' ? JSON.stringify(node.resolved, null, 2) : safeStringify(node.resolved, { indent: 2 });
  }

  const metaProps = computeMetaProps(annotations);
  if (metaProps.length) {
    node.meta = metaProps.join(' ');
  }

  return code(node, _, context);
};

function computeMetaProps(annotations: object) {
  const metaProps = [];
  if (Object.keys(annotations).length) {
    for (const key in annotations) {
      const annotationVal = annotations[key];

      if (typeof annotationVal === 'boolean' || annotationVal === 'true' || annotationVal === 'false') {
        if (annotationVal || annotationVal === 'true') {
          metaProps.push(key);
        }

        // don't add falsey val to meta string
        continue;
      } else if (key === 'type') {
        // this is the only old val we support
        if (annotationVal === 'json_schema') {
          // camelCase to be consistent with rest of annotation props
          metaProps.push('jsonSchema');
        }
      } else if (key === 'highlightLines') {
        // handle deprecated way of adding highlightLines via array annotation
        if (Array.isArray(annotationVal)) {
          const rangeVals = [];
          for (const val of annotationVal) {
            if (Array.isArray(val)) {
              rangeVals.push(`${val[0]}-${val[1]}`);
            } else {
              rangeVals.push(val);
            }
          }
          if (rangeVals.length) {
            metaProps.push(`{${rangeVals.join(',')}}`);
          }
        } else {
          // else we're dealing with the new {1,3} style highlightLines
          metaProps.push(`{${annotationVal}}`);
        }
      } else {
        metaProps.push(`${key}="${annotationVal}"`);
      }
    }
  }

  return metaProps;
}
