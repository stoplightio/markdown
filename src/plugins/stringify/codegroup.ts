import { Handler } from 'mdast-util-to-markdown';
// @ts-expect-error
import flow from 'mdast-util-to-markdown/lib/util/container-flow';

export const codeGroupHandler: Handler = (node, _, context) => {
  const exit = context.enter('codegroup');
  const value = flow(node, context);
  exit();
  return value;
};
