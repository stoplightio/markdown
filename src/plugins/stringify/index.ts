export * from './blockquote';
export * from './code';
export * from './tabs';
