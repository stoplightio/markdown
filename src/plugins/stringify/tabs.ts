import * as Yaml from '@stoplight/yaml';
import { Handler } from 'mdast-util-to-markdown';
// @ts-expect-error
import flow from 'mdast-util-to-markdown/lib/util/container-flow';

const { safeStringify } = Yaml;

export const tabsHandler: Handler = function (node, _, context) {
  const exit = context.enter('tabs');
  const value = flow(node, context);
  exit();
  return `${value}

<!-- type: tab-end -->`;
};

export const tabHandler: Handler = function (node, _, context) {
  const exit = context.enter('tab');

  const { type, ...annotations } = (node.data?.hProperties || {}) as any;

  const value = flow(node, context);

  exit();

  return `<!--
type: tab
${safeStringify(annotations, { skipInvalid: true }).trim()}
-->

${value}`;
};
