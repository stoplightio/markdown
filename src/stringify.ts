import unified, { Node, Processor } from 'unified';
import remarkStringify, { StringifyOpts } from 'remark-stringify';
import remarkParse from 'remark-parse';

const defaultOpts: StringifyOpts = {
  commonmark: true,
  gfm: true,
};

const defaultProcessor = unified()
  .use(remarkParse)
  .use(remarkStringify);

export const stringify = (tree: Node, opts: StringifyOpts = defaultOpts, processor: Processor = defaultProcessor) => {
  return processor
    .data('settings', defaultOpts)
    .stringify(tree);
};
