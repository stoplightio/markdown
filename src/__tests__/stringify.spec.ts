import fs = require('fs');
import path = require('path');
import { stringify } from '../stringify';

describe('stringify', () => {
  it('should work', () => {
    expect(stringify(JSON.parse(fs.readFileSync(path.resolve(__dirname, './fixtures/simple.json'), 'utf-8')))).toMatchSnapshot();
  });
});
