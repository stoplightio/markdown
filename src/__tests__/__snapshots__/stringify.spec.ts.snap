// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringify should work 1`] = `
"**simple**
"
`;
